### PR TITLE
fix(web): avoid re rendering of result item, pagination & poweredBy Image

### PR DESCRIPTION
**Pull request by jyash97**
ReactiveList-Performance

### CODE REVIEW

File: packages/web/src/components/result/addons/Pagination.js [hunk 1]
```diff
@@ -163,6 +164,12 @@ export default function Pagination(props) {
 	);
 }
 
+function areEqual(prevProps, nextProps) {
+	return isEqual(prevProps, nextProps);
+}
+
+export default React.memo(Pagination, areEqual);
```
Comment (bietkul) [inline packages/web/src/components/result/addons/Pagination.js]: React.memo is not supported by React version < 16.6

**Review by bietkul — CHANGES_REQUESTED** [review]

File: packages/web/src/components/result/addons/Pagination.js (hunk 1 above)
Comment (jyash97) [reply packages/web/src/components/result/addons/Pagination.js]: True. 
Should I change to `class` Component and use `PureComponent` ?

Comment (bietkul) [reply packages/web/src/components/result/addons/Pagination.js]: @jyash97 Yes, you can try it out.
Btw what is the reason of extra re-rendering?

Comment (jyash97) [reply packages/web/src/components/result/addons/Pagination.js]: The reason for re-rendering is due to redux store: 

- Promoted Results
- Loading
- Hits Update
- URLParams Update

We can't stop this re-rendering as this can cause some side effects.

**Comment by jyash97** [comment]
Profiler data after applying `PureComponent` and `shouldComponentUpdate` changes:
image

**Comment by jyash97** [comment]
@lakhansamani @bietkul  Is this good to go?

File: packages/web/src/components/result/addons/Pagination.js [hunk 2]
```diff
@@ -65,104 +65,108 @@ const buildPaginationDOM = (props, position) => {
...
+					Next
+				</Button>
+			</div>
+		);
+	}
 }
 
+export default Pagination;
+
 Pagination.propTypes = {
```
Comment (bietkul) [inline packages/web/src/components/result/addons/Pagination.js:168]: @jyash97 Can you put that block before export statement?

File: packages/web/src/components/result/addons/Results.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+Results.propTypes = {
+	hasCustomRender: types.boolRequired,
+	innerClass: types.style,
+	renderItem: types.func,
+	base: types.number,
+	getComponent: types.func,
+	listClass: types.string,
+	triggerClickAnalytics: types.func,
+};
```
Comment (bietkul) [inline packages/web/src/components/result/addons/Results.js:44]: @jyash97 `filteredResults` is not added in props validations, don't you have eslint running?
I'm not sure how we can ignore the eslint errors, we already set a pre-commit hook to avoid the linting errors.
We shouldn't bypass it.

File: packages/web/src/components/result/addons/Results.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import types from '@appbaseio/reactivecore/lib/utils/types';
+
+class Results extends React.Component {
+	shouldComponentUpdate(nextProps) {
+		/*
+			Use shouldComponentUpdate because PureComponent uses shallow comparison of objects.
+			If props contain complex data structures, it may produce false-negatives for
+			deeper differences.
+		*/
+		return !isEqual(nextProps, this.props);
```
Comment (bietkul) [inline packages/web/src/components/result/addons/Results.js]: Would it not create performance issues, suppose you're rendering 50+ results in a page.
It'll be slow to compare props (results) with nested values.
I think PureComponent will be enough.

**Review by bietkul — CHANGES_REQUESTED** [review]

File: packages/web/src/components/result/addons/Results.js (hunk 4 above)
Comment (jyash97) [reply packages/web/src/components/result/addons/Results.js]: I have added the comment for the same do check it. 

I used `PureComponent` component before but it was re-rendering the result.

File: packages/web/src/components/result/addons/Pagination.js (hunk 2 above)
Comment (jyash97) [reply packages/web/src/components/result/addons/Pagination.js:168]: Okay

File: packages/web/src/components/result/addons/Results.js (hunk 3 above)
Comment (jyash97) [reply packages/web/src/components/result/addons/Results.js:44]: Checking

File: packages/web/src/components/result/addons/Results.js (hunk 4 above)
Comment (jyash97) [reply packages/web/src/components/result/addons/Results.js]: Here is the screen of Profiler and console in both scenario for 100 Results:

#### With `shouldComponentUpdate`
image
image


#### With `PureComponent`
image
image

#### With comparing only `filteredResults` in `shouldComponentUpdate`
image
image

File: packages/web/src/components/result/addons/Pagination.js [hunk 5]
```diff
@@ -33,8 +33,8 @@ const buildPaginationDOM = (props, position) => {
 					? start + (Math.ceil(pages / 2) - (pages % 2))
 					: totalPages + 1;
 		}
-		if (currentPage > (totalPages - pages) + 2) {
-			start = (totalPages - pages) + 2;
+		if (currentPage > totalPages - pages + 2) {
+			start = totalPages - pages + 2;
```
Comment (bietkul) [inline packages/web/src/components/result/addons/Pagination.js]: @jyash97 Is your eslint with prettier off, this line has the linting error?
Why there are a lot of unnecessary changes?
Can you update it?

**Review by bietkul — CHANGES_REQUESTED** [review]

File: packages/web/src/components/result/addons/Pagination.js (hunk 5 above)
Comment (jyash97) [reply packages/web/src/components/result/addons/Pagination.js]: There are eslint errors on this file and were there from before I didn't fix them because it would have created more noise in the PR we can do a follow-up PR to fix these errors. They are mainly because of `no-mix-operators`

**Review by bietkul — APPROVED** [review]
